### PR TITLE
feature: Add Prometheus support

### PR DESCRIPTION
**Pull request by pxaws**
# Description of changes
Add Prometheus support in amazon-cloudwatch-agent
* Port existing code for Prometheus support to the opensource repo
* Resolve dependency conflicts
* Introduce metric type handler to add metric type outside Prometheus

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
* All unit and integration tests passed.

### CODE REVIEW

**Review by pingleig — APPROVED** [review]
The license part LGTM w/ the update, though it's better we rewrite that part when we have time... We can fix the style issues later w/ more automated linter etc.

File: plugins/inputs/prometheus_scraper/metrics_type_handler.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+func isInternalMetric(metricName string) bool {
+	//For each endpoint, Prometheus produces a set of internal metrics. See https://prometheus.io/docs/concepts/jobs_instances/
+	if metricName == "up" || strings.HasPrefix(metricName, "scrape_") {
+		return true
+	}
+	return false
+}
+
+// Decorate the Metrics with Metric Types
+func (mf *metricsTypeHandler) Handle(pmb PrometheusMetricBatch) (result PrometheusMetricBatch) {
```
Comment (hdj630) [inline plugins/inputs/prometheus_scraper/metrics_type_handler.go:155]: Be consistent, mf -> mth

File: plugins/inputs/prometheus_scraper/metrics_type_handler.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	// Filter out Summary, Histogram and untyped Metrics and adding logging
+	jobName, instanceId, err := GetScrapeTargetInfo(pmb)
+	if err != nil {
+		log.Printf("E! Failed to get Job Name and Instance ID from Prometheus metrics. \n")
+		return result
+	}
+
+	mc, err := mf.mc.Get(jobName, instanceId)
```
Comment (hdj630) [inline plugins/inputs/prometheus_scraper/metrics_type_handler.go:163]: The variable name "mc" is confusing here, the second "mc" can be renamed as ms which stand for metadataService

File: plugins/inputs/prometheus_scraper/metrics_type_handler.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	}
+	return nil, errors.New("unable to find a target with job=" + job + ", and instance=" + instance)
+}
+
+type ScrapeManager interface {
+	TargetsAll() map[string][]*scrape.Target
+}
+
+type metricsTypeHandler struct {
+	mc *metadataServiceImpl
```
Comment (hdj630) [inline plugins/inputs/prometheus_scraper/metrics_type_handler.go:116]: Why not declare it as interface "*metadataService"?

Comment (pxaws) [reply plugins/inputs/prometheus_scraper/metrics_type_handler.go:116]: Will change it

File: plugins/inputs/prometheus_scraper/metrics_type_handler.go (hunk 2 above)
Comment (pxaws) [reply plugins/inputs/prometheus_scraper/metrics_type_handler.go:163]: Will rename it.

File: plugins/inputs/prometheus_scraper/metrics_type_handler.go (hunk 1 above)
Comment (pxaws) [reply plugins/inputs/prometheus_scraper/metrics_type_handler.go:155]: Will rename it

File: go.mod [hunk 4]
```diff
@@ -3,35 +3,61 @@ module github.com/aws/amazon-cloudwatch-agent
...
 	github.com/aws/aws-sdk-go v1.30.15
 	github.com/bigkevmcd/go-configparser v0.0.0-20200217161103-d137835d2579
 	github.com/docker/docker v1.13.1
+	github.com/go-kit/kit v0.10.0
 	github.com/gobwas/glob v0.2.3
 	github.com/google/cadvisor v0.36.0
+	github.com/hashicorp/golang-lru v0.5.4
 	github.com/imdario/mergo v0.3.8 // indirect
-	github.com/influxdata/telegraf v1.15.2
+	github.com/influxdata/telegraf v0.0.0-00010101000000-000000000000
```
[the remark this answers is not in the document]
Comment (pxaws) [reply go.mod:38]: The telegraf version doesn't matter here because we do the replacement above.

**Review by hdj630 — APPROVED** [review]

**Pull request merged**